main/medium/src/reaper.rs

```diff
@@ -5573,6 +5748,461 @@ impl<UsageScope> Reaper<UsageScope> {
         NonNull::new(ptr).ok_or(ReaperFunctionError::new("couldn't get MIDI editor take"))
     }
 
+    /// Create a new MIDI media item, containing no MIDI events.
+    ///
+    /// Time is in seconds unless time_in_qn is true.
+    ///
+    /// # Safety
+    ///
+    /// REAPER can crash if you pass an invalid track.
+    pub unsafe fn create_midi_item_in_proj(
+        &self,
+        track: MediaTrack,
+        starttime: f64,
+        endtime: f64,
+        time_in_qn: bool,
```

Comment:
Taking a bool is not very self-explaining in this case (see [here](https://github.com/helgoboss/reaper-rs/blob/b7878f758721e661531609434a891f2ac540a129/main/medium/src/lib.rs#L66-L66) for design rationale):

```rust
reaper.create_midi_item_in_proj(track, 0.0, 10.0, true);
```

I think I would prefer this one:

```rust
reaper.create_midi_item_in_proj(track, PositionInQuarterNotes(0.0)..PositionInQuarterNotes(10.0));
```

This would also make sense because we already have newtypes `PositionInQuarterNotes` and `PositionInSeconds` which are used consistently across the medium-level API. In real-world use cases, chances are you already have `PositionInQuarterNotes` values ready, so this would get quite terse.

It should be achievable by having:

```rust
pub unsafe fn create_midi_item_in_proj(&self, track: MediaTrack, range: R) -> ReaperFunctionResult<MediaItem>
where
        R: Into<TimeRange>,
        UsageScope: MainThreadOnly,
{
    ...
}

pub enum TimeRange {
    Seconds(Range<PositionInSeconds>),
    QuarterNotes(Range<PositionInQuarterNotes>),
}

impl From<Range<PositionInSeconds>> for TimeRange {
    ...
}

impl From<Range<PositionInQuarterNotes>> for TimeRange {
    ...
}

```

Alternatively, we could also start by taking a `TimeRange` directly, without the magic `Into` conversions (and add the magic later if it's desired):

```rust
reaper.create_midi_item_in_proj(track, TimeRange::QuarterNotes(PositionInQuarterNotes(0.0)..PositionInQuarterNotes(10.0)));
```

But a bit too much repetition for my taste.

BTW, Rust has inclusive and exclusive ranges. Although it wouldn't make any different for us because we don't want to iterate the range, I think it would be good to communicate the intention. If I understand REAPER correctly, the end time always represents something exclusive. The item actually ends before that end position. What do you think?

Reply:
Well, in this case we again face the problem of what medium-level should do, and what not...

For my personal tase, I found, that `std` already has awesome type, that solve 90% of ambiguity and boiler-plate: `std:time::Duration`. While one think of a Position in the therm of «Duration from the start», it removes any leaks in terminology. Alongside, it leaves interpretation of low-level API f64 arguments not on the type system, but on concrete wrapper-function realization. And, if, occasionally, someone made a mistake in interpreting `float time` as seconds, while it is not — the change will not affect next-level users.

More that, this implements accurate arithmetic operations as position comparison and holding the length of events. It naturally converts from\into f64, it is fast and supports all the derive traits, like `Copy, Hash, Ord, Eq` etc.

Reply:
The same thing I thought, when used a `Color`. NativeColor and RGB color is just an leaking abstraction, that not really helps in code usage. So, again, for myself I've made just `Color` with `from_native` and `into_native` methods, that take and return f64.